src/server/routes/entity/entity.tsx

```diff
@@ -1312,3 +1341,186 @@ export function displayPreview(req:PassportRequest, res:$Response, next) {
+				authorId: editorJSON.id,
+				isMerge: false
+			}).save(null, {transacting});
+			const additionalProps = _.pick(entityForm, additionalEntityProps[_.snakeCase(entityType)]);
+			const changedProps = await getChangedProps(
+				orm, transacting, true, currentEntity, entityForm, entityType,
+				newRevision, additionalProps
+			);
+			const mainEntity = await fetchOrCreateMainEntity(
+				orm, transacting, true, currentEntity.bbid, entityType
+			);
+			mainEntity.shouldInsert = true;
+
+			// set changed attributes on main entity
+			_.forOwn(changedProps, (value, key) => mainEntity.set(key, value));
+			const savedMainEntity = await saveEntitiesAndFinishRevision(
+				orm, transacting, true, newRevision, mainEntity, [mainEntity],
+				editorJSON.id, entityForm.note
+			);
+
+			/* We need to load the aliases for search reindexing and refresh it*/
+			await savedMainEntity.load('aliasSet.aliases', {transacting});
+
+			/* New entities will lack some attributes like 'type' required for search indexing */
+			await savedMainEntity.refresh({transacting});
+
+			/* fetch and reindex EditionGroups that may have been created automatically by the ORM and not indexed */
+			if (savedMainEntity.get('type') === 'Edition') {
+				await indexAutoCreatedEditionGroup(orm, savedMainEntity, transacting);
+			}
+			bbidMap[entityKey] = savedMainEntity.get('bbid');
+			savedMainEntities[entityKey] = savedMainEntity.toJSON();
+		}
+		try {
+			// bookshelf's transaction have issue with Promise.All
+			await Object.keys(body).reduce((promise, entityKey) => promise.then(() => processEntity(entityKey)), Promise.resolve());
+
+			// adding relationship on newly created entites
+			await Promise.all(Object.keys(allRelationships).map(async (entityId) => {
+				const rels = allRelationships[entityId];
+				if (!_.isEmpty(rels)) {
+					const relationships = rels.map((rel) => (
+						{...rel, sourceBbid: _.get(bbidMap, rel.sourceBbid) ?? rel.sourceBbid,
+						 targetBbid: _.get(bbidMap, rel.targetBbid) ?? rel.targetBbid}
+					));
+					const mainEntity = savedMainEntities[entityId];
+					const {relationshipSetId} = await handleAddRelationship({relationships}, editorJSON.id,
+						 mainEntity, mainEntity.type, orm, transacting);
+					mainEntity.relationshipSetId = relationshipSetId;
+				}
+			}));
+			return savedMainEntities;
+		}
+		catch (err) {
+			log.error(err);
+			throw err;
+		}
+	});
+	const achievementPromise = entityEditPromise.then(
+		(entitiesJSON:Record<string, any>) => {
+			const entitiesAchievementsPromise = [];
+			for (const entityJSON of Object.values(entitiesJSON)) {
+				entitiesAchievementsPromise.push(achievement.processEdit(
+					orm, editorJSON.id, entityJSON.revisionId
+				)
+					.then((unlock) => {
+						if (unlock.alert) {
+							entityJSON.alert = unlock.alert;
+						}
+						return entityJSON;
+					}));
+			}
+			return Promise.all(entitiesAchievementsPromise).catch(err => { throw err; });
+		}
+	);
```

Comment:
Now that we're duplicating this code, this could do with 1. Being refactored into a separate reusable function and 2. Being refactored to use async/await syntax